File: solutions/4118-design-ride-sharing-system/solution.py
class RideSharingSystem:

    def __init__(self):
        self.riders = deque()
        self.drivers = deque()
        self.cancelled = set()

    def addRider(self, riderId: int) -> None:
        self.riders.append(riderId)

    def addDriver(self, driverId: int) -> None:
        self.drivers.append(driverId)

    def matchDriverWithRider(self) -> List[int]:

        while self.riders and self.riders[0] in self.cancelled:
            self.cancelled.discard(self.riders.popleft())

        if self.riders and self.drivers:
            d = self.drivers.popleft()
            r = self.riders.popleft()
            return d, r
        return (-1,-1)


    def cancelRider(self, riderId: int) -> None:
        if riderId in self.riders:
            self.cancelled.add(riderId)

'''

class RideSharingSystem:

    def __init__(self):
        self.riders = deque()
        self.drivers = deque()
        self.cancelled = set()

    def addRider(self, riderId: int) -> None:
        self.riders.append(riderId)

    def addDriver(self, driverId: int) -> None:
        self.drivers.append(driverId)

    def matchDriverWithRider(self) -> List[int]:
        # Skip cancelled riders at the front
        while self.riders and self.riders[0] in self.cancelled:
            self.cancelled.discard(self.riders.popleft())

        if self.riders and self.drivers:
            r = self.riders.popleft()
            d = self.drivers.popleft()
            return [d, r]
        return [-1, -1]

    def cancelRider(self, riderId: int) -> None:
        self.cancelled.add(riderId)  # O(1) instead of O(n)

'''

File: solutions/4118-design-ride-sharing-system/solution.cpp
class RideSharingSystem {
public:
deque<int> riders;
deque<int> drivers;

RideSharingSystem() {}

void addRider(int riderId) {
riders.push_back(riderId);
}

void addDriver(int driverId) {
drivers.push_back(driverId);
}

vector<int> matchDriverWithRider() {
if (!riders.empty() && !drivers.empty()) {
int d = drivers.front();
int r = riders.front();
drivers.pop_front();
riders.pop_front();
return {d, r};
}
return {-1, -1};
}

void cancelRider(int riderId) {
for (int i = 0; i < (int)riders.size(); ++i) {
if (riders[i] == riderId) {
riders.erase(riders.begin() + i);
break;
}
}
}
};
